### ISSUE TITLE
[Discussion](NPU) Gradient Mismatch in KLDivLoss: Native implementation skips target gradients.

### ISSUE THREAD

**Issue by kiritorl** [linkedin/Liger-Kernel#1021]
I encountered a gradient mismatch issue while validating the test_jsd.py. The native implementation appears to treat the target as a constant and skips gradient computation, even when requires_grad=True is set.
```
import torch
from torch.nn import KLDivLoss
log_m = torch.tensor([[-1.2584, -0.5437, -2.0000],
                      [-0.6250, -1.6068, -1.3282]], device="npu", requires_grad=True)
log_q_resize = torch.tensor([[-2.1719, -0.2139, -2.5469],
                             [-0.6328, -3.4062, -0.8281]], device="npu", requires_grad=True)
kl = KLDivLoss(reduction="none", log_target=True)
out = kl(log_m, log_q_resize)
# out = torch.exp(log_q_resize) * (log_q_resize - log_m)
out.backward(torch.ones_like(out))
print(log_m.grad)
print(log_q_resize.grad)  # npu return none
```

Should we replace the native implementation with a composed torch implementation as the test baseline?

**Comment by Tcc0403** [comment]
~We don't have to calculate gradient of `target` because `target` is ground truth which is supposed to be treated as a constant.  Is `target.grad` checked somewhere in our tests?~

My bad, in jsd, `target` is a mixture of p and q, so gradients of `target` should also be considered.
